### PR TITLE
New api

### LINKED ISSUE #22675: Request to get all product images
#### Is your feature request related to a problem?

Drugin synchronization of my system Products with Prestashop, Ineed to trigger request for each product to get product images. As a result, I am sending lots of requests like that:
https://presta:8888/api/images/products/1096
https://presta:8888/api/images/products/1097
... and so on
Thats very resource wasteful

#### Describe the solution you'd like

Simple request to get all of product images, forparsing later:
https://presta:8888/api/images/products
should return XML (or json) of all product images, where I can see which image is related to which product

#### Alternatives you've considered

Alternative is to get one product a time, which I said is very resource wasteful

### PR DESCRIPTION
**Pull request by KminekMatej**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | See below, its pretty huge
| Type?             | bug fix & improvement & new feature & refactor
| Category?         | WS
| BC breaks?        | yes
| Deprecations?     | no
| Fixed ticket?     | Fixes #22675
| How to test?      | Using new WS autotest
| Possible impacts? | Composing API has been completely refactored, modernized and enlighted. Every file underneath classes/webservice has been changed, some now added. There is just one actual impact, which has brought me to refactor API completely - getting product images. Please see diff in ([ https://www.diffchecker.com/fisRdJ1l ],) what has been actually changed. This should be the only BC break, contained in this PR, everything else should stay the same (as for now)

**Note**:
This PR replaces pull-request #23100

**Description**:

WebService API was having historical troubles. Its XML output has been using string composing of headers / footers (instead of using modern SimpleXML PHP interface), which could cause troubles (well actually didnt, at least in modern versions of Prestashop). Troubles started after JSON output has been added into Prestashop - you can't correctly compose JSON output using header and footer rendering. That led to some custom overriding functions, leading to a bit of chaos on API Core structure.

I tried to solve this design trouble by creating a class called ApiNode and refactored WebServiceOutputBuilder to firstly build structure of ApiNode tree. Each ApiNode is basically an object, composing of name, type, array of attributes and array of subnodes. This ApiNode is then processed by XML / JSON rendering classes.

Then, WebServiceOutputJSON and WebServiceOutputXML are using simple function renderNode(ApiNode), which iterates over ApiNode (and its children) and composes output string.

WebServiceOutputXML uses SuperXMLElement (which is actually SimpleXMLElement, just extended with addCData() function) to create XML hierarchy. Output is then properly XML formatted into string. That makes XML API output easily extendable in the future. XML output is no more pretty printed (using deprecated $depth property), to avoid unneccessary overhead (XML is usually pretty printed by modern browsers automatically anyway and there is no need for pretty printing in actual API calls)

WebServiceOutputJSON iterates over ApiNode and its children to create a simple array of objects and its subobjects, then transforms it into string, using json_encode common PHP function. Due to nature of JSON, we cannot put attributes in it, as in XML. Either we have to accept complete BC break and add attributes into JSON request as well, or avoid working with attributes and treat important attributes as children nodes, thus creating another BC break in XML output.
(Working with attributes will definitely be neccessary on images requests - but they havent been working anyway in JSON, so I believe there is no BC break). Non-BC possibility would thus be to add attributes as subnodes only on Images requests. But this solution needs further analyze.

Consider this PR as a draft. I am going to process further tests. I have already tested few requests (mainly on products), ensuring that XML and JSON outputs stays the same, but more tests are neccessary. I am not sure whether there are any API autotests, which would come very handy.

Also, code-review from original creators of Prestashop WebService API would be very handy. I think that during the code examination I discovered every potential use-case, but one can never be sure.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27297)

### CODE REVIEW

**Comment by Hlavtox** [comment]
Borec

File: classes/webservice/ApiNode.php [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+ */
+class ApiNode
+{
+    public const TYPE_VALUE = 'value';
+    public const TYPE_LANGUAGE = 'language';
+    public const TYPE_PARENT = 'parent';
+    public const TYPE_LIST = 'list';
+
+    public static $languages;
+    private string $type;
```
Comment (PierreRambaud) [inline classes/webservice/ApiNode.php]: You can't do it with PHP7.2

File: classes/webservice/ApiNode.php [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+        }
+
+        if (!empty($field['synopsis_details']) && $schemaToDisplay !== 'blank') {
+            foreach ($field['synopsis_details'] as $name => $detail) {
+                $newNode->addAttribute($name, is_array($detail) ? implode(' ', $detail) : $detail);
+            }
+        }
+
+        // display i18n fields
+        if (isset($field['i18n']) && $field['i18n']) {
```
Comment (PierreRambaud) [inline classes/webservice/ApiNode.php]:
```suggestion
        if (!empty($field['i18n'])) {
```

File: classes/webservice/WebserviceOutputBuilder.php [hunk 3]
```diff
@@ -278,25 +285,28 @@ public function setStatus($num)
...
-                    }
+        if (empty($errors)) {
+            return '';
+        }
+
+        if (!isset($this->objectRender)) {
+            return '<pre>' . print_r($errors, true) . '</pre>';
+        }
+
+        $this->output = ApiNode::list('errors');
```
Comment (PierreRambaud) [inline classes/webservice/WebserviceOutputBuilder.php:296]: Do we really need to use Singleton?

Comment (KminekMatej) [reply classes/webservice/WebserviceOutputBuilder.php:296]: Its not actually singleton pattern, just a shortcut to utilize multiple constructors. You could replace that with Factory or with simple constructor with fluent setters.
Factory would need another class.
Constructor with fluent setters gets more error-prone for future development, in my opinion.

File: classes/webservice/WebserviceOutputBuilder.php [hunk 4]
```diff
@@ -659,79 +609,109 @@ protected function renderAssociations($object, $depth, $associations, $ws_params
...
+    /**
+     * Inject one association section into parent node
+     *
+     * @param ApiNode $parentNode
+     * @param string $assoc_name
+     * @param array $params
+     *
+     * @return void
+     */
+    private function injectAssociation(&$parentNode, $object, $assoc_name, $association, $params, $class_name, $objects_assoc, $fields_assoc)
```
Comment (KminekMatej) [inline classes/webservice/WebserviceOutputBuilder.php]: I'm bit unsure here. My imlpementation hereis not using $class_name parameter in here, which could lead to errors. Could please someone point to a situation where $class_name has been used, in previous version - so I can test it and upgrade it?

**Comment by KminekMatej** [comment]
Is it really neccessary to check there are no merge commits?

**Comment by matks** [comment]
> Is it really neccessary to check there are no merge commits?

Yes, it makes git history harder to manage so it makes the task for maintainers harder.

It's much easier to understand what's going on in a linear history.

**Comment by KminekMatej** [comment]
@matks So that means I should create new branch and apply all the changes again onto it? Or how do you solve it, usually?

**Comment by matks** [comment]
> @matks So that means I should create new branch and apply all the changes again onto it? Or how do you solve it, usually?

I suggest

- if you are not familiar with this, first create a backup of your work
- checkout `develop` branch up-to-date
- create new branch using `develop` as base point
- use [git cherry-pick](https://git-scm.com/docs/git-cherry-pick) to apply the commits of your PR (just the ones with your changes, not the merge commits) to new branch

**Comment by KminekMatej** [comment]
@matks Thats a very clever way to go, thanks. Done that, so please follow PR #27366

**Pull request closed without merging**